### ISSUE TITLE
Rate limiting does not function properly

### ISSUE THREAD

**Issue by acd124** [chrisdewa/unbelipy#7]
Opening a new issue this time as there have been updates and the other title would not make as much sense.

1. Making several identical requests with retry_rate_limits=True and prevent_rate_limits=True results in the client waiting a full 5 seconds between every single request, rather than only once all 10 of the requests per bucket are used up.
2. The rate limit code shows no evidence of differentiating the rate limits for a request to a guild 305129477627969547 and for a guild 264445053596991498, as far as I can tell just lumping all requests under the get_balance method under the same limit.

**Comment by chrisdewa** [comment]
Yes, I'm still working on those. I thought the rate limits applied on a method basis (get guild, patch guild, etc) but as you pointed out they're guild and user specific i need to rework them. I need to run tests and when I'm done I'll update.

Again thanks for bringing this up.

**Comment by acd124** [comment]
In my example I specifically only mentioned different guilds, different users do not change the rate limit bucket.

**Comment by chrisdewa** [comment]
pull request #8  improves rate limit handling but i still need to run tests with concurrent different buckets.
retry_rate_limits is still being held until preventing rate limits is done.
I'll leave the issue open until then. For now, it should be safe to test

**Comment by acd124** [comment]
Cannot currently test since python and pip are telling me that version 1.1.0b does not exist.

**Comment by chrisdewa** [comment]
yeah, sorry, forgot to publish (it was already too late). Should be up now

**Comment by acd124** [comment]
The get_leaderboard method created a bucket called `GET/guilds/441117219162816512/users?offset=1`. Parameters like offset do not create another bucket. Additionally, though this is not strictly necessary, I think that there should be a slightly better distinction between when there is a user id after the /users/ and when there is not after the /users.

**Comment by acd124** [comment]
client.retry_rate_limits currently waits the retry_after given by the the api, but that number of seconds, rather than milliseconds, I did not wait the full duration because I do not want to waste 2 hours with the code just waiting.

**Comment by acd124** [comment]
client.prevent_rate_limits doesn't give you the result for the api request that gets the bucket to 0 until the bucket has reset. I don't want to wait 8 seconds for the results when you already have it. The waiting should be done by the request that is made when the stored bucket remaining is 0, not when the received bucket remaining is 0.

**Comment by chrisdewa** [comment]
Yes, it happens because the throttler is set on _aexit__ but it should happen on __aenter__ I'll patch later today.

**Comment by chrisdewa** [comment]
commit 93b36d5 from version 1.1.1b should fix this later issue. If everything is good, ill move to working on ratelimit retrying

**Comment by chrisdewa** [comment]
commit c1927ed should finally fix this issue

**Comment by acd124** [comment]
This looks fixed
